### PR TITLE
Include a built-in Finch HTTP adapter

### PR DESCRIPTION
**Pull request by andrewhr**
Following the approach of [Swoosh](https://hexdocs.pm/swoosh/Swoosh.ApiClient.Finch.html), include a Finch-based adapter. The patch keeps hackney adapter as default given it doesn't require any extra setup.

In this patch the implementation of Hackney adapter got moved, which should not be a breaking change for most cases - unless someone is explicit referring to hackney implementation. The intend behind moving the impl is to help make it clear there are many impls and keep it separate from the behaviour definition.

A small helper to update `http_client` key is also included, with the intention to ease usage in cases where any create helper is used.

Example:

```elixir
   AWS.Client.create()
   |> AWS.Client.put_http_client({AWS.HTTPClient.Finch, []})
   |> AWS.S3.list_buckets()
```

Using Application environment was considered as alternative to the helper, but goes against what most of this library does.

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -32,6 +38,7 @@ defmodule AWS.Mixfile do
       {:earmark, "~> 1.4", only: [:dev]},
       {:ex_doc, "~> 0.24", only: [:dev]},
       {:bypass, "~> 2.1", only: [:test]},
+      {:finch, "~> 0.14.0 or ~> 0.14.0", optional: true},
```
Comment (philss) [inline mix.exs:41]: Shouldn't this be `~> 0.13.0 or ~> 0.14.0`:

```suggestion
      {:finch, "~> 0.13.0 or ~> 0.14.0", optional: true},
```

**Review by philss — APPROVED** [review]
Awesome! :D

Just a minor thing and :shipit:

File: mix.exs (hunk 1 above)
Comment (andrewhr) [reply mix.exs:41]: Yep, sorry 🤦

(Btw, the compatibility goes beyond that but I kept these just to keep it simple yet friendly to upgrades)

**Comment by philss** [comment]
@andrewhr thank you! :purple_heart:

**Pull request merged**